blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/AbstractElasticIO.java

```diff
@@ -0,0 +1,444 @@
+            resultList.add(new TokenInfo(entry.getKey(), entry.getValue()));
+        }
+        return resultList;
+    }
+
+
+    /**
+     * Performs terms aggregation by metric_name which returns doc_count by
+     * metric_name index that matches the given regex.
+     *
+     *  Sample request body:
+     *
+     *  {
+     *      "size": 0,
+     *      "query": {
+     *          "term": {
+     *              "tenantId": "836986"
+     *          }
+     *      },
+     *      "aggs": {
+     *          "metric_name_tokens": {
+     *              "terms": {
+     *                  "field" : "metric_name",
+     *                  "include": "<regex>",
+     *                  "size": 0
+     *              }
+     *          }
+     *      }
+     *  }
+     *
+     * @param tenant
+     * @param regexMetricName
+     * @return
+     */
+    private SearchResponse aggregateByMetricTokens(final String tenant, final String regexMetricName) {
+
+        AggregationBuilder aggregationBuilder =
+                AggregationBuilders.terms(METRICS_TOKENS_AGGREGATE)
+                        .field(ESFieldLabel.metric_name.name())
+                        .include(regexMetricName)
+                        .size(0);
+
+        return client.prepareSearch(new String[] {ELASTICSEARCH_INDEX_NAME_READ})
+                .setRouting(tenant)
+                .setSize(0)
+                .setVersion(true)
+                .setQuery(QueryBuilders.termQuery(ESFieldLabel.tenantId.toString(), tenant))
+                .addAggregation(aggregationBuilder)
+                .execute()
+                .actionGet();
+    }
+
+
+    /**
+     * From a given terms aggregate results, it identifies the following kind of tokens
+     *
+     * 1) Next level of tokens for a given prefix which also have subsequent next level (not checking for enums yet)
+     * 2) For the metric names which do not fall under above category, identifies if any of them are complete metric
+     *    names at a given prefix level(LEVEL_0) or at a next level(LEVEL_1).
+     *
+     * @param aggregateTerms
+     * @param totalTokensInPrefix
+     * @param tokensWithNextLevel  Next level of tokens for a given prefix which also have subsequent next level
+     * @param completeMetricNamesByLevel contains complete metric names.
+     */
+    private void processMetricTokensAggregate(final Terms aggregateTerms,
+                                              final int totalTokensInPrefix,
+                                              final Set<String> tokensWithNextLevel,
+                                              final Map<Integer, Set<String>> completeMetricNamesByLevel) {
+
+        //0 indicates ground level. if prefix is foo.bar, 0 is at foo.bar level
+        final Map<String, Long> partialMetrics0LevelMap = new LinkedHashMap<String, Long>();
+        final Map<String, Long> partialMetrics0LevelAggMap = new LinkedHashMap<String, Long>();
```

Comment:
Can you explain what's exactly held in these maps?  I don't understand the difference between *Map & *AggMap